data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java

```diff
@@ -508,21 +512,103 @@ public JacksonEvent build() {
+            } else if (getExcludeKeys() != null && !getExcludeKeys().isEmpty()) {
+                jsonString = searchAndFilter(baseNode, "", getExcludeKeys(), EXCLUDE_ALL);
+            } else {
+                jsonString = baseNode.toString();
+            }
+
             final String tagsKey = getTagsKey();
-            if(tagsKey != null) {
+            if (tagsKey != null) {
                 final JsonNode tagsNode = mapper.valueToTree(event.getMetadata().getTags());
-                return jsonString.substring(0, jsonString.length()-1) + ",\""+tagsKey+"\":" + tagsNode.toString()+"}";
+                return jsonString.substring(0, jsonString.length() - 1) + ",\"" + tagsKey + "\":" + tagsNode.toString() + "}";
             }
             return jsonString;
         }
+
+        /**
+         * Perform DFS(Depth-first search) like traversing using recursion on the Json Tree and return the json string.
+         * This supports filtering (to include or exclude) from a list of keys.
+         *
+         * @param node         Root node to start traversing
+         * @param path         Json path, e.g. /foo/bar
+         * @param filterKeys   A list of filtered keys
+         * @param filterAction Either to include (RETAIN_ALL or true) or to exclude (EXCLUDE_ALL or false)
+         * @return a json string with filtered keys
+         */
+        private String searchAndFilter(JsonNode node, String path, final List<String> filterKeys, boolean filterAction) {
+
+            if (node.isArray()) { // for array node.
+                StringJoiner sj = new StringJoiner(",", "[", "]");
+                node.forEach(childNode -> sj.add(searchAndFilter(childNode, path, filterKeys, filterAction)));
+                return sj.toString();
+            } else {
+                StringJoiner sj = new StringJoiner(",", "{", "}");
+                List<String> valueList = new ArrayList<>();
+
+                node.properties().forEach(entry -> {
+                    String keyPath = path + SEPARATOR + entry.getKey();
+                    // Track whether the key is found in the filter list.
+                    // Different behaviours between include and exclude action.
+                    boolean found = false;
+                    for (String key : filterKeys) {
+                        if (keyPath.equals(key)) {
+                            found = true;
+                            // To keep the order.
+                            if (filterAction == RETAIN_ALL) {
+                                valueList.add("\"" + entry.getKey() + "\":" + entry.getValue().toString());
+                            }
+                            break;
+                        } else if (key.startsWith(keyPath)) {
+                            found = true;
+                            valueList.add("\"" + entry.getKey() + "\":" + searchAndFilter(entry.getValue(), keyPath, filterKeys, filterAction));
```

Comment:
Is there any reason to prefer the recursive or non-recursive stack based implementation of DFS here?

Reply:
Non-recursive requires an extra use of stack in memory but it's not a problem. I think both ways have the same Os, it's just recursive is more easier to read and implement, especially for array nodes. Let me know if any concerns here.

Reply:
The depth of these events will probably not be deep enough to make stack overflow an issue. I think this is acceptable.